Render stacktraces properly

aritra24 opened the pull request:
Closes: #47921

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

aritra24 committed:
Render stacktraces properly
Closes: #47921

bbovenzi commented on airflow/ui/src/components/RenderStructuredLog.tsx:
`renderStructuredLog` since it doesn't actually return a named component

aritra24 replied:
Ack

bbovenzi commented on airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -16,22 +16,17 @@
 import dayjs from "dayjs";
 import innerText from "react-innertext";
-import { Link } from "react-router-dom";
+import type { JSX } from "react/jsx-runtime";
```
We don't need this import

aritra24 replied:
Ack

jscheffl commented on airflow/ui/src/components/RenderStructuredLog.tsx:
```diff
@@ -0,0 +1,177 @@
+      );
+    }
+  }
+
```
Proposal alternative to render:
```suggestion
      return (
        <details key={error.exc_type} open={true} style={{ marginLeft: "20em" }}>
          <summary data-testid={`summary-${error.exc_type}`}>
            <chakra.span color="fg.info" cursor="pointer">
              {error.exc_type}: {error.exc_value}
            </chakra.span>
          </summary>
          {errorLines}
        </details>
      );
    });
  }

  elements.push(
    <chakra.span className="event" key={2} style={{ whiteSpace: "pre-wrap" }}>
      {event}:
    </chakra.span>,
  );

  for (const key in structured) {
    if (Object.hasOwn(structured, key)) {
      elements.push(
        " ",
        <chakra.span color={key === "logger" ? "fg.info" : undefined} key={`prop_${key}`}>
          {key === "logger" ? "source" : key}={JSON.stringify(structured[key])}
        </chakra.span>,
      );
    }
  }

  elements.push(
    <chakra.span className="event" key={2} style={{ whiteSpace: "pre-wrap" }}>
      {details}
    </chakra.span>,
  );

```

Then it would look like:
![image](https://github.com/user-attachments/assets/599da19d-6ae2-419a-b15b-273f69a46d4d)

...and folded:
![image](https://github.com/user-attachments/assets/2d6d2a76-95fd-4969-b8a1-88ab6ac1e2ce)

aritra24 committed:
Update airflow/ui/src/components/RenderStructuredLog.tsx

Co-authored-by: Jens Scheffler <95105677+jscheffl@users.noreply.github.com>

bbovenzi reviewed (approved):
Jens' suggestion looks like the right UX. Let's just rebase

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/RenderStructuredLog.tsx:
```diff
@@ -0,0 +1,180 @@
+        <chakra.p key="test">
+          File{" "}
+          <chakra.span color="fg.info" key="test">
+            &quot;{frame.filename}&quot;
```
Not sure we need to encode there, this seem to also work
```suggestion
            "{frame.filename}"
```

aritra24 replied:
I think the linter had complained against it but I'll recheck.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/RenderStructuredLog.tsx:
```diff
@@ -0,0 +1,180 @@
+  let details;
+
+  if (errorDetail !== undefined) {
+    delete structured.error_detail;
```
Not fan of the delete, we could spread and use that:

```js
const { errorDetail, ...structuredWithoutDetails} = profiles;
```

`structuredWithoutDetails` is not a good name but it's just for the example.

aritra24 replied:
Ack, fair I'll update it.

pierrejeambrun reviewed (approved):
Nice addition 🎉 

Just a couple of nits, non blocking

aritra24 committed:
Fixes ut, resolves comments

Fixed the UT which was failing due to a
mistaken id, and removed the usage of the
quotes as escaped characters and usage of delete

jscheffl commented on airflow-core/src/airflow/ui/src/queries/useLogs.tsx:
```diff
@@ -16,22 +16,17 @@
 import dayjs from "dayjs";
 import innerText from "react-innertext";
-import { Link } from "react-router-dom";
+import type { JSX } from "react/jsx-runtime";
```
Import really not needed. Can you remove? Then I think it is good to merge.
```suggestion
```

aritra24 replied:
Oops, I think it's auto adding it somehow. I removed it a couple times. It somehow sneaks back in if I don't pay attention to the auto imports.

aritra24 committed:
Remove unsed import

Co-authored-by: Jens Scheffler <95105677+jscheffl@users.noreply.github.com>

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/renderStructuredLog.tsx:
```diff
@@ -0,0 +1,180 @@
+  let details;
+
+  if (errorDetail !== undefined) {
+    delete structured.error_detail;
```
You don't need that anymore since we use `reStructured` from now on.

aritra24 committed:
Merge branch 'main' into fix_logging

pierrejeambrun committed:
Remove delete statement
